### PR TITLE
[GLUTEN-3425] Create not existing Hdfs folder in gluten side when writing hdfs file.

### PR DESCRIPTION
**Pull request by JkSelf**
## What changes were proposed in this pull request?

When Gluten calls the Velox Parquet writer to write a Parquet file, the temporary path obtained from Spark may not have been created yet. While writing a local file, Velox automatically creates the necessary file path if it does not exist. However, this is not the case for HDFS paths. We attempted to create the HDFS path on the Velox side, but the community was not receptive to this approach. As a result, we decided to create the HDFS path on the Gluten side.

## How was this patch tested?

manual local test

### CODE REVIEW

**Comment by JkSelf** [comment]
@rui-mo Can you help to review? Thanks.

**Review by rui-mo — COMMENTED** [review]
Can we add a test to ensure the functionality?

**Review by rui-mo — DISMISSED** [review]

File: ep/build-velox/src/get_velox.sh [hunk 1]
```diff
@@ -16,8 +16,8 @@
 
 set -exu
 
-VELOX_REPO=https://github.com/oap-project/velox.git
-VELOX_BRANCH=main
+VELOX_REPO=https://github.com/JkSelf/velox.git
+VELOX_BRANCH=craete-write-path
```
Comment (rui-mo) [inline ep/build-velox/src/get_velox.sh:20]: Remove?

**Review by rui-mo — APPROVED** [review]

File: backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxFormatWriterInjects.scala [hunk 2]
```diff
@@ -47,6 +47,15 @@ trait VeloxFormatWriterInjects extends GlutenFormatWriterInjectsBase {
       nativeConf: java.util.Map[String, String]): OutputWriter = {
     val originPath = path
 
+    // Created the hdfs path if not existed.
```
Comment (rui-mo) [inline backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxFormatWriterInjects.scala:50]: Created -> Create?

File: backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxFormatWriterInjects.scala [hunk 3]
```diff
@@ -47,6 +47,15 @@ trait VeloxFormatWriterInjects extends GlutenFormatWriterInjectsBase {
...
+    if (path.substring(0, 5).equals("hdfs:")) {
```
Comment (rui-mo) [inline backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxFormatWriterInjects.scala:51]: Maybe declare `hdfs:` as a const variable, and use its length in substring.

Comment (JkSelf) [reply backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxFormatWriterInjects.scala:52]: @rui-mo Updated.

File: backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxFormatWriterInjects.scala (hunk 2 above)
Comment (JkSelf) [reply backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxFormatWriterInjects.scala]: @rui-mo Updated.

File: backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxFormatWriterInjects.scala [hunk 4]
```diff
@@ -47,6 +47,16 @@ trait VeloxFormatWriterInjects extends GlutenFormatWriterInjectsBase {
       nativeConf: java.util.Map[String, String]): OutputWriter = {
     val originPath = path
 
+    // Create the hdfs path if not existed.
+    val hdfsPrefix = "hdfs:"
```
Comment (PHILO-HE) [inline backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxFormatWriterInjects.scala:51]: Nit:
I think it is better to use: `val hdfsScheme = "hdfs://"`

File: backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxFormatWriterInjects.scala (hunk 3 above)
Comment (PHILO-HE) [reply backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxFormatWriterInjects.scala]: Nit:
`path.startsWith(hdfsScheme)`

BTW, not related to this PR: `originPath` looks redundant in this part of code. Can we just use `path` instead?`?

**Review by PHILO-HE — COMMENTED** [review]
LGTM! Just few trivial comment.

File: backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxFormatWriterInjects.scala (hunk 4 above)
Comment (JkSelf) [reply backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxFormatWriterInjects.scala]: @PHILO-HE Updated.

File: backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxFormatWriterInjects.scala [hunk 5]
```diff
@@ -45,8 +45,20 @@ trait VeloxFormatWriterInjects extends GlutenFormatWriterInjectsBase {
       dataSchema: StructType,
       context: TaskAttemptContext,
       nativeConf: java.util.Map[String, String]): OutputWriter = {
+    // The path parameter is same with the method name of OutputWriter.
```
Comment (JkSelf) [inline backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxFormatWriterInjects.scala:48]: @PHILO-HE Add comments for why need to rename the path parameter to originPath.

**Comment by JkSelf** [comment]
> Can we add a test to ensure the functionality?

Offline discussion with @rui-mo . It is hard to add the unit test to test.

File: backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxFormatWriterInjects.scala (hunk 5 above)
Comment (PHILO-HE) [reply backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxFormatWriterInjects.scala:48]: Thanks for your feedback! Can we just use a no conflict name for argument in `createOutputWriter(path: String..`?

Comment (JkSelf) [reply backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxFormatWriterInjects.scala]: @PHILO-HE Updated. Thanks for your suggestions.

**Review by PHILO-HE — APPROVED** [review]
LGTM! Thanks!

**Comment by zhouyuan** [comment]
@JkSelf 
The parquet write job failed on Jenkins with recent Velox changes: 
```
Job aborted due to stage failure: Task 0 in stage 2.0 failed 4 times, most recent failure: Lost task 0.3 in stage 2.0 (TID 2000) (sr270 executor 9): java.lang.RuntimeException: Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Failed to open hdfs file: /user/sparkuser/ETL/newparquet_zstd/_temporary/0/_temporary/attempt_202310190834151922475248029547116_0002_m_000000_2000/part-00000-e52c3743-6c6b-459e-b6c6-a2bc676fa493-c000.zstd.parquet, with error: /user/sparkuser/ETL/newparquet_zstd/_temporary/0/_temporary/attempt_202310190834151922475248029547116_0002_m_000000_2000/part-00000-e52c3743-6c6b-459e-b6c6-a2bc676fa493-c000.zstd.parquet already exists as a directory
```

**Comment by PHILO-HE** [comment]
> @JkSelf The parquet write job failed on Jenkins with recent Velox changes:
> 
> ```
> Job aborted due to stage failure: Task 0 in stage 2.0 failed 4 times, most recent failure: Lost task 0.3 in stage 2.0 (TID 2000) (sr270 executor 9): java.lang.RuntimeException: Exception: VeloxRuntimeError
> Error Source: RUNTIME
> Error Code: INVALID_STATE
> Reason: Failed to open hdfs file: /user/sparkuser/ETL/newparquet_zstd/_temporary/0/_temporary/attempt_202310190834151922475248029547116_0002_m_000000_2000/part-00000-e52c3743-6c6b-459e-b6c6-a2bc676fa493-c000.zstd.parquet, with error: /user/sparkuser/ETL/newparquet_zstd/_temporary/0/_temporary/attempt_202310190834151922475248029547116_0002_m_000000_2000/part-00000-e52c3743-6c6b-459e-b6c6-a2bc676fa493-c000.zstd.parquet already exists as a directory
> ```

Looks we should make parent directory extracted from `filePath`?

**Comment by JkSelf** [comment]
> > @JkSelf The parquet write job failed on Jenkins with recent Velox changes:
> > ```
> > Job aborted due to stage failure: Task 0 in stage 2.0 failed 4 times, most recent failure: Lost task 0.3 in stage 2.0 (TID 2000) (sr270 executor 9): java.lang.RuntimeException: Exception: VeloxRuntimeError
> > Error Source: RUNTIME
> > Error Code: INVALID_STATE
> > Reason: Failed to open hdfs file: /user/sparkuser/ETL/newparquet_zstd/_temporary/0/_temporary/attempt_202310190834151922475248029547116_0002_m_000000_2000/part-00000-e52c3743-6c6b-459e-b6c6-a2bc676fa493-c000.zstd.parquet, with error: /user/sparkuser/ETL/newparquet_zstd/_temporary/0/_temporary/attempt_202310190834151922475248029547116_0002_m_000000_2000/part-00000-e52c3743-6c6b-459e-b6c6-a2bc676fa493-c000.zstd.parquet already exists as a directory
> > ```
> 
> Looks we should make parent directory extracted from `filePath`?

It is my mistake to use the wrong test scripts and caused this error not found previously. Yes. It should be the parent directory. I use the parent directory and got the following exception in velox. It seems we can't create the path in gluten. I may need some time to investigate the root cause. And  will revert this PR firstly. 

```

2023-10-20 07:31:10,288 ERROR util.TaskResources: Task 25 failed by error:
java.lang.RuntimeException: IOError: Couldn't serialize thrift: Insufficient space in external MemoryBuffer

        at io.glutenproject.datasource.DatasourceJniWrapper.close(Native Method)
        at org.apache.spark.sql.execution.datasources.velox.VeloxFormatWriterInjects$$anon$1.close(VeloxFormatWriterInjects.scala:117)
        at org.apache.spark.sql.execution.datasources.FileFormatDataWriter.releaseCurrentWriter(FileFormatDataWriter.scala:75)
        at org.apache.spark.sql.execution.datasources.FileFormatDataWriter.releaseResources(FileFormatDataWriter.scala:86)
        at org.apache.spark.sql.execution.datasources.FileFormatDataWriter.commit(FileFormatDataWriter.scala:116)
        at org.apache.spark.sql.execution.datasources.FileFormatWriter$.$anonfun$executeTask$1(FileFormatWriter.scala:386)
        at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1525)
        at org.apache.spark.sql.execution.datasources.FileFormatWriter$.executeTask(FileFormatWriter.scala:394)
        at org.apache.spark.sql.execution.datasources.FileFormatWriter$.$anonfun$write$17(FileFormatWriter.scala:303)
        at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
        at org.apache.spark.scheduler.Task.run(Task.scala:131)
        at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1491)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
2023-10-20 07:31:10,293 ERROR datasources.FileFormatWriter: Job job_202310200731034252066363461590636_0025 aborted.
I1020 07:31:10.298817 796153 FileSink.cpp:123] closing file: hdfs://sr246:9000/user/sparkuser/parquet-write-1g/_temporary/0/_temporary/attempt_202310200731034252066363461590636_0025_m_000000_25/part-00000-289d2c68-b0ca-4e8d-8877-59250b390ced-c000.zstd.parquet,  total size: 137.07MB
E1020 07:31:10.462782 796153 Exceptions.h:69] Line: ../../velox/connectors/hive/storage_adapters/hdfs/HdfsWriteFile.cpp:57, Function:close, Expression: success == 0 (-1 vs. 0) Failed to close hdfs file: File does not exist: /user/sparkuser/parquet-write-1g/_temporary/0/_temporary/attempt_202310200731034252066363461590636_0025_m_000000_25/part-00000-289d2c68-b0ca-4e8d-8877-59250b390ced-c000.zstd.parquet (inode 0) Holder libhdfs3_client_rand_0.124599_count_1_pid_795698_tid_140455000884992 does not have any open files.
        at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.checkLease(FSNamesystem.java:2880)
        at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.fsync(FSNamesystem.java:3352)
        at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.fsync(NameNodeRpcServer.java:1438)
        at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.fsync(ClientNamenodeProtocolServerSideTranslatorPB.java:1049)
        at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java)
        at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:524)
        at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1025)
        at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:876)
        at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:822)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.Subject.doAs(Subject.java:422)
        at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1730)
        at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2682)
, Source: RUNTIME, ErrorCode: INVALID_STATE

```

**Pull request merged**